### PR TITLE
HttpWebRequest PNSE/NotImplemented members implemented & code coverage increased

### PR DESCRIPTION
**Pull request by ViktorHofer**
Fixes https://github.com/dotnet/corefx/issues/18680
Fixes https://github.com/dotnet/corefx/issues/18900
Fixes https://github.com/dotnet/corefx/issues/18800

Also refactored code in HttpWebRequest.
Added a lot of tests which should raise the code coverage to 80-90%

### CODE REVIEW

File: src/System.Net.Requests/src/System/Net/HttpWebRequest.cs [hunk 1]
```diff
@@ -1097,20 +1003,26 @@ private Task<Stream> GetRequestStreamTask()
...
                     request.Content = new ByteArrayContent(bytes.Array, bytes.Offset, bytes.Count);
                 }
-
                 handler.AutomaticDecompression = AutomaticDecompression;
-                handler.Credentials = _credentials;
+                handler.Credentials = Credentials;
                 handler.AllowAutoRedirect = AllowAutoRedirect;
                 handler.MaxAutomaticRedirections = MaximumAutomaticRedirections;
-                handler.MaxResponseHeadersLength = MaximumResponseHeadersLength;
+                handler.MaxResponseHeadersLength = MaximumResponseHeadersLength <= 0 ?
```
Comment (ViktorHofer) [inline src/System.Net.Requests/src/System/Net/HttpWebRequest.cs]: Please review especially this line and check if that is what we want.

File: src/System.Net.Requests/src/Resources/Strings.resx [hunk 2]
```diff
@@ -265,4 +265,7 @@
   <data name="net_ftpstatuscode_ActionNotTakenFilenameNotAllowed" xml:space="preserve">
     <value>File name not allowed</value>
   </data>
-</root>
+  <data name="net_invalid_host" xml:space="preserve">
+    <value>The specified value is not a valid Host header string.</value>
+  </data>
+</root>
```
Comment (davidsh) [inline src/System.Net.Requests/src/Resources/Strings.resx]: nit: missing newline at end of line

File: src/System.Net.WebClient/tests/WebClientTest.cs [hunk 3]
```diff
@@ -399,8 +399,8 @@ public static void UseDefaultCredentials_Roundtrips()
             await Assert.ThrowsAsync<WebException>(() => wc.DownloadStringTaskAsync(System.Net.Test.Common.Configuration.Http.RemoteEchoServer));
         }
 
-        [ActiveIssue(18680)]
-        [OuterLoop("Networking test talking to remote server: issue #11345")]
+        //[ActiveIssue(18680)]
+        //[OuterLoop("Networking test talking to remote server: issue #11345")]
```
Comment (davidsh) [inline src/System.Net.WebClient/tests/WebClientTest.cs]: If you are fixing the bug, then you should delete the ActiveIssue.  But you need to keep the Outerloop one.  You shouldn't leave them as commented out.

Comment (ViktorHofer) [inline src/System.Net.WebClient/tests/WebClientTest.cs]: ah sry that was solely for testing purposes. Forgot that

File: src/System.Net.Requests/src/System/Net/HttpWebRequest.cs [hunk 4]
```diff
@@ -1097,20 +1003,29 @@ private Task<Stream> GetRequestStreamTask()
...
                     Threading.Timeout.InfiniteTimeSpan :
                     TimeSpan.FromMilliseconds(Timeout);
+                if (_hostUri != null)
+                {
+                    client.DefaultRequestHeaders.Host = _hostUri.Host;
+                }
+                if (!string.IsNullOrEmpty(MediaType))
+                {
+                    client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue(MediaType));
+                }
```
Comment (ViktorHofer) [inline src/System.Net.Requests/src/System/Net/HttpWebRequest.cs]: And check if these two headers are set correctly pls

**Comment by davidsh** [comment]
Thanks for your PR and for helping networking!

I have some comments. I think your PR here should be split up.  The issue you are addressing is #18680 which is about WebClient only and only for a few properties needing parameter validation.

But this PR is much more involved than that.  You are adding functionality to both HttpWebRequest and WebClient that goes beyond just parameter validation.

So, please split up this PR.  There should be a simpler PR that just addresses #18680.  Then perhaps separate PRs that add all of this functionality for HttpWebRequest properties, etc.  We also need to consider that the additional PRs are much more complex and possibly we might defer them post 2.0 due to risk.

cc: @karelz @stephentoub

File: src/System.Net.Requests/src/System/Net/HttpWebRequest.cs (hunk 4 above)
Comment (ViktorHofer) [inline src/System.Net.Requests/src/System/Net/HttpWebRequest.cs]: MediaType is wrong

**Comment by davidsh** [comment]
Also, please run outerloop for  Windows and *Nix when changing networking code.  You probably just have to run Windows and Ubuntu outerloop which will give us enough CI test coverage.

**Comment by ViktorHofer** [comment]
Let us discuss the PR shortly. The issue is about WebClient only, that's true, but it uses the HttpWebRequest behind it and the properties there weren't implemented. 

The other changes aren't behavior changes they are only involve code refactoring (auto properties instead of variables, simplified body expressions). 

Of course the diff tool shows lot of changes but as I just tried to point out, most of them are only refactoring (which don't change any behavior...). I'm still open to split them but I don't think that it's necessary for this PR.

**Comment by davidsh** [comment]
In general, we try to keep PRs as focused as possible especially as we near a release.  So, "refactoring" changes like this should be done in a separate PR.

Also, with all the added functionality appearing in this PR, there doesn't seem to be an equal amount of new test coverage being added. So, with the new/changed properties, how are they being tested to make sure the changes are valid?

**Comment by ViktorHofer** [comment]
You are absolutely right. I will do that!

**Comment by ViktorHofer** [comment]
Added tests. Code coverage should be far better now.

@karelz Yeah I agree that it makes sense.

**Comment by ViktorHofer** [comment]
Now I need to divide the code for different PRs *ufff*

**Comment by karelz** [comment]
Thanks! I hope it won't be too much of a challenge ;-)

**Comment by ViktorHofer** [comment]
Removed refactoring code. PTAL @davidsh

**Comment by ViktorHofer** [comment]
@karelz @davidsh When this gets merged I will push the refactoring PR

File: src/System.Net.Requests/src/System/Net/HttpWebRequest.cs [hunk 5]
```diff
@@ -335,45 +383,46 @@ public string TransferEncoding
...
 
-                //
-                // if not check if the user is trying to set chunked:
-                //
-                string newValue = value.ToLower();
-                fChunked = (newValue.IndexOf(ChunkedHeader) != -1);
-
-                //
-                // prevent them from adding chunked, or from adding an Encoding without
-                //  turing on chunked, the reason is due to the HTTP Spec which prevents
```
Comment (davidsh) [inline src/System.Net.Requests/src/System/Net/HttpWebRequest.cs]: spelling: "turning"

File: src/System.Net.Requests/src/System/Net/HttpWebRequest.cs [hunk 6]
```diff
@@ -1158,7 +1214,16 @@ private Task<Stream> GetRequestStreamTask()
                             request.Content = new ByteArrayContent(Array.Empty<byte>());
                         }
 
-                        request.Content.Headers.TryAddWithoutValidation(headerName, _webHeaderCollection[headerName]);
+                        // Add MediaType to ContentType seperated by a semicolon
```
Comment (davidsh) [inline src/System.Net.Requests/src/System/Net/HttpWebRequest.cs]: spelling: "separated"

**Comment by davidsh** [comment]
@dotnet-bot test outerloop netcoreapp Windows_NT Debug
@dotnet-bot test outerloop netcoreapp Ubuntu16.04 Debug

File: src/System.Net.Requests/src/System/Net/HttpWebRequest.cs [hunk 7]
```diff
@@ -45,13 +46,18 @@ public class HttpWebRequest : WebRequest, ISerializable
...
         private int _endGetResponseCalled = 0;
-                
+
         private int _maximumAllowedRedirections = HttpHandlerDefaults.DefaultMaxAutomaticRedirections;
         private int _maximumResponseHeadersLen = _defaultMaxResponseHeadersLength;
         private ServicePoint _servicePoint;
         private int _timeout = WebRequest.DefaultTimeoutMilliseconds;
         private HttpContinueDelegate _continueDelegate;
 
+        // stores the user provided Host header as Uri. If the user specified a default port explicitly we'll loose
```
Comment (danmosemsft) [inline src/System.Net.Requests/src/System/Net/HttpWebRequest.cs]: nit : "loose" -> "lose"

File: src/System.Net.Requests/src/System/Net/HttpWebRequest.cs [hunk 8]
```diff
@@ -269,11 +280,54 @@ public string Host
...
+                    throw new ArgumentNullException();
+                }
+
+                Uri hostUri;
+                if ((value.IndexOf('/') != -1) || (!TryGetHostUri(value, out hostUri)))
+                {
+                    throw new ArgumentException(SR.net_invalid_host);
+                }
+
+                // ** Requires CAS **
```
Comment (danmosemsft) [inline src/System.Net.Requests/src/System/Net/HttpWebRequest.cs]: can you just remove these two comment lines

**Comment by ViktorHofer** [comment]
Fixed spelling and added a lot more tests. 18 properties still need code coverage (~40 tests) to be over 90% coverage (estimation).

**Comment by ViktorHofer** [comment]
Added more tests. Code coverage should now be at maximum *yeah*. 

With that, my code refactoring changes - which I will submit in a separate PR - are 100% covered.

**Comment by ViktorHofer** [comment]
@davidsh I'd like to bring that PR forward. In regards to #18900, can you please take another look?

**Comment by davidsh** [comment]
cc: @karelz 
@stephentoub Can you also take another look thru this?  There are a lot of changes (over a dozen commits some of which revert previous commits).

Wonder if it would be better if this got rebased and squashed so that we can review fresh?

**Comment by ViktorHofer** [comment]
@davidsh @stephentoub Rebase and squash done

**Comment by karelz** [comment]
@ViktorHofer I would personally keep real changes (new code & code modifications) separate from formatting changes (spaces, removed comments, formatting, etc.). At least separate them in different commits.
I know it is tempting to make formatting changes when you're already editing the file (I often have the same urge). However, given how close we are to ZBB, I think we should try to avoid such mixing of changes to ease it up on code reviews and lower the risk.
Anyway, let's keep this PR as it is now - just a thought for future.

Also, I know this may be border-line personal opinion, so maybe others will disagree with me - in which case feel free to ignore my suggestions :)

**Comment by ViktorHofer** [comment]
@karelz Thanks for the feedback! I didn't had the intention to make formating changes but unfortunately when I removed my refactoring changes (and overwrote the file with the latest version from master, per copy and paste) Visual Studio auto-formatted the code and yeah... I'll try to avoid it the next time.

**Comment by ViktorHofer** [comment]
ping @stephentoub

File: src/System.Net.Requests/src/System/Net/HttpWebRequest.cs [hunk 9]
```diff
@@ -143,21 +149,26 @@ public virtual bool AllowReadStreamBuffering
...
+            get => _maximumResponseHeadersLen;
             set
             {
+                if (RequestSubmitted)
+                {
+                    throw new InvalidOperationException(SR.net_reqsubmitted);
+                }
+                if (value < 0 && value != System.Threading.Timeout.Infinite)
+                {
+                    throw new ArgumentOutOfRangeException("value", SR.net_toosmall);
```
Comment (stephentoub) [inline src/System.Net.Requests/src/System/Net/HttpWebRequest.cs]: Nit: name of(value)

File: src/System.Net.Requests/src/System/Net/HttpWebRequest.cs [hunk 10]
```diff
@@ -269,11 +280,51 @@ public string Host
...
             set
             {
-                throw new PlatformNotSupportedException();
+                if (RequestSubmitted)
+                {
+                    throw new InvalidOperationException(SR.net_writestarted);
+                }
+                if (value == null)
+                {
+                    throw new ArgumentNullException();
```
Comment (stephentoub) [inline src/System.Net.Requests/src/System/Net/HttpWebRequest.cs]: Nit: nameof(value)

File: src/System.Net.Requests/src/System/Net/HttpWebRequest.cs [hunk 11]
```diff
@@ -269,11 +280,51 @@ public string Host
         {
             get
             {
-                throw new PlatformNotSupportedException();
+                Uri hostUri = _hostUri ?? Address;
+                return (_hostUri == null || !_hostHasPort) && Address.IsDefaultPort ?
+                    hostUri.Host :
+                    $"{hostUri.Host}:{hostUri.Port}";
```
Comment (stephentoub) [inline src/System.Net.Requests/src/System/Net/HttpWebRequest.cs]: hostUri.Host + ":" + hostUri.Port is more efficient

File: src/System.Net.Requests/src/System/Net/HttpWebRequest.cs [hunk 12]
```diff
@@ -1464,5 +1519,17 @@ private static string DateToString(DateTime D)
...
+        {
+            // Use StringBuilder rather than UriBuilder, since we don't know if the "hostName" will contain a
+            // port or not.
+            StringBuilder sb = new StringBuilder(Address.Scheme);
+            sb.Append("://");
+            sb.Append(hostName);
+            sb.Append(Address.PathAndQuery);
+
+            return Uri.TryCreate(sb.ToString(), UriKind.Absolute, out hostUri);
+        }
```
Comment (stephentoub) [inline src/System.Net.Requests/src/System/Net/HttpWebRequest.cs]: Are we already including StringBuilderCache in this project? If so, we should use it here.

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs [hunk 13]
```diff
@@ -122,6 +128,30 @@ public void AllowReadStreamBuffering_SetTrueThenGet_ExpectTrue(Uri remoteServer)
         }
 
         [Theory, MemberData(nameof(EchoServers))]
+        public void ContentLength_SetAfterRequestSubmitted_ThrowsInvalidOperationException(Uri remoteServer)
+        {
+            HttpWebRequest request = WebRequest.CreateHttp(remoteServer);
+            request.GetResponse();
```
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: Let's dispose of everything that's disposable in tests like this.

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs [hunk 14]
```diff
@@ -161,6 +240,430 @@ public void ContinueTimeout_SetNegativeTwo_ThrowsArgumentOutOfRangeException(Uri
...
+        {
+            // Save default value
+            int defaultMaximumErrorsResponseLength = HttpWebRequest.DefaultMaximumErrorResponseLength;
+
+            int newDefaultMaximumErrorsResponseLength = 255;
+            HttpWebRequest.DefaultMaximumErrorResponseLength = newDefaultMaximumErrorsResponseLength;
+            Assert.Equal(newDefaultMaximumErrorsResponseLength, HttpWebRequest.DefaultMaximumErrorResponseLength);
+
+            // Cleanup
+            HttpWebRequest.DefaultMaximumErrorResponseLength = defaultMaximumErrorsResponseLength;
```
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: Reset code like this would ideally be in a finally.

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs [hunk 15]
```diff
@@ -161,6 +240,430 @@ public void ContinueTimeout_SetNegativeTwo_ThrowsArgumentOutOfRangeException(Uri
...
+            HttpWebRequest.DefaultMaximumResponseHeadersLength = defaultMaximumResponseHeadersLength;
+        }
+
+        [Theory, MemberData(nameof(EchoServers))]
+        public void DefaultMaximumErrorResponseLength_SetAndGetLength_ValuesMatch(Uri remoteServer)
+        {
+            // Save default value
+            int defaultMaximumErrorsResponseLength = HttpWebRequest.DefaultMaximumErrorResponseLength;
+
+            int newDefaultMaximumErrorsResponseLength = 255;
```
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: const?

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs [hunk 16]
```diff
@@ -161,6 +240,430 @@ public void ContinueTimeout_SetNegativeTwo_ThrowsArgumentOutOfRangeException(Uri
...
+            HttpWebRequest.DefaultMaximumErrorResponseLength = newDefaultMaximumErrorsResponseLength;
```
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: Is this going to cause any problems for any tests that might run concurrently? Or all tests in this project serialized?

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs [hunk 17]
```diff
@@ -161,6 +240,430 @@ public void ContinueTimeout_SetNegativeTwo_ThrowsArgumentOutOfRangeException(Uri
...
+            var serverCertificateVallidationCallback = new Security.RemoteCertificateValidationCallback((a, b, c, d) => true);
+            request.ServerCertificateValidationCallback = serverCertificateVallidationCallback;
+            Assert.Same(serverCertificateVallidationCallback, request.ServerCertificateValidationCallback);
+        }
+
+        [Theory, MemberData(nameof(EchoServers))]
+        public void ClientCertificates_SetNullX509_ThrowsArgumentNullException(Uri remoteServer)
+        {
+            HttpWebRequest request = WebRequest.CreateHttp(remoteServer);
+            Assert.Throws<ArgumentNullException>(() => request.ClientCertificates = null);
```
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: Can we include parameter name validation in such throws checks?

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs [hunk 18]
```diff
@@ -521,10 +1051,13 @@ public void Abort_BeginGetRequestStreamThenAbort_EndGetRequestStreamThrowsWebExc
 
             request.BeginGetResponse(new AsyncCallback(RequestStreamCallback), state);
 
-            request.Abort();
-            Assert.Equal(1, state.RequestStreamCallbackCallCount);
-            WebException wex = state.SavedRequestStreamException as WebException;
-            Assert.Equal(WebExceptionStatus.RequestCanceled, wex.Status);
+            if (!request.HaveResponse)
```
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: Are there problematic race conditions with this pattern? Seems like either there would be our this check shouldn't be necessary.

File: src/System.Net.Requests/src/System/Net/HttpWebRequest.cs (hunk 11 above)
Comment (ViktorHofer) [inline src/System.Net.Requests/src/System/Net/HttpWebRequest.cs]: I guess the most performant way without creating a side string would be with StringBuilder?

File: src/System.Net.Requests/src/System/Net/HttpWebRequest.cs (hunk 12 above)
Comment (ViktorHofer) [inline src/System.Net.Requests/src/System/Net/HttpWebRequest.cs]: Not in the project yet

File: src/System.Net.Requests/src/System/Net/HttpWebRequest.cs (hunk 11 above)
Comment (stephentoub) [inline src/System.Net.Requests/src/System/Net/HttpWebRequest.cs]: No, in this case with string.Concat, which is what's generated for the code I showed

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs (hunk 16 above)
Comment (ViktorHofer) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: When I the static tests I hoped that they would run serialized... But I didn't check. Where to find that?

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs (hunk 18 above)
Comment (ViktorHofer) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: Ya there are race conditions and possible deadlocks. This is by design according to MSDN. There is an issue for these tests here.

**Comment by ViktorHofer** [comment]
@stephentoub Addressed your feedback. PTAL

**Comment by davidsh** [comment]
@dotnet-bot test outerloop netcoreapp Windows_NT Debug
@dotnet-bot test outerloop netcoreapp Ubuntu16.04 Debug

**Comment by davidsh** [comment]
@ViktorHofer Pls always run outerloop test when changing networking code.  Most networking tests are outerrloop.

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs (hunk 18 above)
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: > Ya there are race conditions and possible deadlocks.

We should not add tests that could deadlock or non-deterministically fail.

File: src/System.Net.Requests/src/System/Net/HttpWebRequest.cs (hunk 12 above)
Comment (stephentoub) [inline src/System.Net.Requests/src/System/Net/HttpWebRequest.cs]: Ok.  This would be better anyway as:
```C#
string s = Address.Scheme + "://" + hostname + Address.PathAndQuery;
return Uri.TryCreate(s, UriKind.Absolute, out hostUri);
```

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs (hunk 18 above)
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: What is the "by design" according to MSDN?  What could deadlock here?

There's certainly a race condition with regards to whether the request will have completed by the time Abort is called; nothing to do about that.  But that just means the checks for whether the operation completes successfully or not should accommodate the potential of the request succeeding or being canceled.  Adding the check you've added only decreases the chances of hitting the race condition, but doesn't remove it entirely.

And to do this actually correctly, rather than going against a remote server we don't control, we should really be going against a local loopback server such that we can guarantee the abort will be requested before the operation has completed.

Comment (ViktorHofer) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: These are the methods which are problematic:

```
(deactivated) Abort_BeginGetRequestStreamThenAbort_EndGetRequestStreamThrowsWebException(Uri remoteServer)
Abort_BeginGetResponseThenAbort_ResponseCallbackCalledBeforeAbortReturns(Uri remoteServer)
(deactivated) Abort_BeginGetResponseThenAbort_EndGetResponseThrowsWebException(Uri remoteServer)
Abort_BeginGetResponseUsingNoCallbackThenAbort_Success(Uri remoteServer)
Abort_CreateRequestThenAbort_Success(Uri remoteServer)
```

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs (hunk 16 above)
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: > When I the static tests I hoped that they would run serialized

Whether test methods are static or instance doesn't affect whether they're serialized.

> Where to find that?

By default xunit will serialize all tests on the same test class, but it will allow tests from different classes to run in parallel.  To override that and disable parallelization for all tests in a test assembly, an assembly-level attribute can be applied, e.g.
https://github.com/dotnet/corefx/blob/master/src/System.Linq.Parallel/tests/XunitAssemblyAttributes.cs#L10
I don't see one in this test assembly, which suggests tests from different test classes in this assembly can run concurrently.

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs (hunk 18 above)
Comment (ViktorHofer) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: See Remarks in https://msdn.microsoft.com/en-US/library/system.net.httpwebrequest.abort(v=vs.110).aspx

Is seems like Abort was always broken in netfx. Not sure if this is still true with the current corefx implementation.

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs (hunk 16 above)
Comment (ViktorHofer) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: Great, thanks for the info! Still I think we should be safe as we aren't touching any parts of HttpWebRequest in other tests classes.

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs (hunk 18 above)
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: It's not saying they inherently deadlock.  It's saying you need to be aware of the behavior so that you don't write code that deadlocks, just being aware that Abort may cause the callback to be synchronously invoked, so you shouldn't write code with dependencies between that callback and the caller that would result in a deadlock.

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs (hunk 16 above)
Comment (ViktorHofer) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: Ah not true :/ I just saw that there is another class which tests static members: `HttpWebRequestHeaderTest.cs`

Comment (ViktorHofer) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: We could merge the tests into HttpWebRequest (which would make sense IMO).

**Comment by ViktorHofer** [comment]
@dotnet-bot test outerloop netcoreapp Windows_NT Debug
@dotnet-bot test outerloop netcoreapp Ubuntu16.04 Debug

**Comment by ViktorHofer** [comment]
@stephentoub Changed Abort tests to use loopback server. PTAL --> I hope we can sign it off now so that it gets in before 5/10

**Comment by ViktorHofer** [comment]
@dotnet-bot test outerloop netcoreapp Windows_NT Debug
@dotnet-bot test outerloop netcoreapp Ubuntu16.04 Debug

**Comment by ViktorHofer** [comment]
@dotnet-bot test outerloop netcoreapp Windows_NT Debug

File: src/System.Net.Requests/src/System/Net/HttpWebRequest.cs [hunk 19]
```diff
@@ -269,11 +280,51 @@ public string Host
...
+                {
+                    int endOfIPv6Address = value.IndexOf(']');
+                    if (endOfIPv6Address == -1)
+                    {
+                        _hostHasPort = true;
+                    }
+                    else
+                    {
+                        _hostHasPort = value.LastIndexOf(':') > endOfIPv6Address;
+                    }
```
Comment (stephentoub) [inline src/System.Net.Requests/src/System/Net/HttpWebRequest.cs]: Nit: this could be simplified to:
```C#
int endOfIPv6Address = value.IndexOf(']');
_hostHasPort = endOfIPv6Address == -1 || value.LastIndexOf(':') > endOfIPv6Address;
```

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs [hunk 20]
```diff
@@ -157,7 +255,545 @@ public void ContinueTimeout_SetNegativeOne_Success(Uri remoteServer)
...
+
+                var sw = Stopwatch.StartNew();
+                WebException exception = Assert.Throws<WebException>(() =>
+                {
+                    var response = (HttpWebResponse)request.GetResponse();
+                });
+                
+                sw.Stop();
+
+                Assert.InRange(sw.ElapsedMilliseconds, 1, 5 * 1000);
```
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: Why did you change the timeout from https://github.com/dotnet/corefx/pull/19118/files#diff-c6a038cf9c04ca8b1fbd38a930477f91L214?

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs [hunk 21]
```diff
@@ -157,7 +255,545 @@ public void ContinueTimeout_SetNegativeOne_Success(Uri remoteServer)
...
+        public void Address_CtorAddress_ValuesMatch(Uri remoteServer)
+        {
+            HttpWebRequest request = WebRequest.CreateHttp(remoteServer);
+            Assert.Equal(remoteServer, request.Address);
+        }
+
+        [Theory, MemberData(nameof(EchoServers))]
+        public void UserAgent_SetThenGetWindows_ValuesMatch(Uri remoteServer)
+        {
+            const string userAgent = "Windows";
```
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: Nit: UserAgent

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs [hunk 22]
```diff
@@ -157,7 +255,545 @@ public void ContinueTimeout_SetNegativeOne_Success(Uri remoteServer)
...
         {
             HttpWebRequest request = WebRequest.CreateHttp(remoteServer);
-            Assert.Throws<ArgumentOutOfRangeException>(() => request.ContinueTimeout = -2);
+            Assert.Throws<ArgumentOutOfRangeException>("value", () => request.ContinueTimeout = -2);
+        }
+
+        [Theory, MemberData(nameof(EchoServers))]
+        public void Timeout_SetThenGetZero_ExpectZero(Uri remoteServer)
+        {
+            const int timeout = 0;
```
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: Nit: Timeout

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs [hunk 23]
```diff
@@ -139,6 +184,59 @@ public void ContentType_SetThenGetEmptyValue_ExpectNull(Uri remoteServer)
...
+        {
+            HttpWebRequest request = WebRequest.CreateHttp(remoteServer);
+            Assert.Throws<ArgumentException>("value", () => request.MaximumAutomaticRedirections = 0);
+            Assert.Throws<ArgumentException>("value", () => request.MaximumAutomaticRedirections = -1);
+        }
+
+        [Theory, MemberData(nameof(EchoServers))]
+        public void MaximumAutomaticRedirections_SetThenGetOne_Success(Uri remoteServer)
+        {
+            const int maximumAutomaticRedirections = 1;
```
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: Nit: MaximumAutomaticRedirections

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs [hunk 24]
```diff
@@ -139,6 +184,59 @@ public void ContentType_SetThenGetEmptyValue_ExpectNull(Uri remoteServer)
...
+        public void MaximumResponseHeadersLength_SetNegativeTwo_ThrowsArgumentOutOfRangeException(Uri remoteServer)
+        {
+            HttpWebRequest request = WebRequest.CreateHttp(remoteServer);
+            Assert.Throws<ArgumentOutOfRangeException>("value", () => request.MaximumResponseHeadersLength = -2);
+        }
+
+        [Theory, MemberData(nameof(EchoServers))]
+        public void MaximumResponseHeadersLength_SetThenGetNegativeOne_Success(Uri remoteServer)
+        {
+            const int maximumResponseHeaderLength = -1;
```
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: Nit: MaximumResponseHeaderLength

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs [hunk 25]
```diff
@@ -110,15 +130,40 @@ public void AllowReadStreamBuffering_SetTrueThenGet_ExpectTrue(Uri remoteServer)
...
+        public void ContentLength_SetNegativeOne_ThrowsArgumentOutOfRangeException(Uri remoteServer)
+        {
+            HttpWebRequest request = WebRequest.CreateHttp(remoteServer);
+            Assert.Throws<ArgumentOutOfRangeException>("value", () => request.ContentLength = -1);
+        }
+
+        [Theory, MemberData(nameof(EchoServers))]
+        public void ContentLength_SetThenGetOne_Success(Uri remoteServer)
+        {
+            const int contentLength = 1;
```
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: Nit: ContentLength

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs (hunk 20 above)
Comment (ViktorHofer) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: It was set to that value because an external remote endpoint was used. Now it uses a local loopback server

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs [hunk 26]
```diff
@@ -26,26 +29,43 @@ public partial class HttpWebRequestTest
         public HttpWebRequestTest(ITestOutputHelper output)
         {
             _output = output;
+
+            if (PlatformDetection.IsFullFramework)
```
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: Do we need to special-case this?

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs [hunk 27]
```diff
@@ -157,7 +255,545 @@ public void ContinueTimeout_SetNegativeOne_Success(Uri remoteServer)
...
+            request.Pipelined = true;
+            Assert.True(request.Pipelined);
+            request.Pipelined = false;
+            Assert.False(request.Pipelined);
+        }
+
+        [Theory, MemberData(nameof(EchoServers))]
+        public void Referer_SetThenGetReferer_ValuesMatch(Uri remoteServer)
+        {
+            const string referer = "Referer";
```
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: Nit: Referer

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs [hunk 28]
```diff
@@ -157,7 +255,545 @@ public void ContinueTimeout_SetNegativeOne_Success(Uri remoteServer)
...
+            HttpWebRequest request = WebRequest.CreateHttp(remoteServer);
+            request.Referer = referer;
+            Assert.Equal(referer, request.Referer);
+        }
+
+        [Theory, MemberData(nameof(EchoServers))]
+        public void TransferEncoding_NullOrWhiteSpace_ValuesMatch(Uri remoteServer)
+        {
+            const string transferEncoding = "xml";
```
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: Nit: TransferEncoding

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs [hunk 29]
```diff
@@ -157,7 +255,545 @@ public void ContinueTimeout_SetNegativeOne_Success(Uri remoteServer)
...
+            using (var response = (HttpWebResponse)request.GetResponse())
+            {
+                Assert.True(request.PreAuthenticate);
+            }
+        }
+
+        [Theory, MemberData(nameof(EchoServers))]
+        public void Connection_NullOrWhiteSpace_ValuesMatch(Uri remoteServer)
+        {
+            const string connection = "connect";
```
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: Nit: Connection (I'll stop commenting on these const casings)

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs [hunk 30]
```diff
@@ -157,7 +255,545 @@ public void ContinueTimeout_SetNegativeOne_Success(Uri remoteServer)
...
+        {
+            HttpWebRequest request = WebRequest.CreateHttp(remoteServer);
+            Assert.Throws<ArgumentException>("value", () => request.Expect = "100-continue");
+        }
+
+        [Theory, MemberData(nameof(EchoServers))]
+        public void DefaultMaximumResponseHeadersLength_SetAndGetLength_ValuesMatch(Uri remoteServer)
+        {
+            int defaultMaximumResponseHeadersLength = HttpWebRequest.DefaultMaximumResponseHeadersLength;
+            int newDefaultMaximumResponseHeadersLength = 255;
```
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: Nit: const?

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs (hunk 20 above)
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: I see.  It may still be too small; under load in CI servers things can take longer than you'd otherwise expect on rare occasions.

**Comment by stephentoub** [comment]
Please squash when merging. Thanks.

**Comment by ViktorHofer** [comment]
@dotnet-bot test outerloop netcoreapp Windows_NT Debug
@dotnet-bot test outerloop netcoreapp Ubuntu16.04 Debug
